README.md
This server connects to a heating pump control system made by the company Hanazeder and reads sensor measurements, outlet / pump states and energy values.
This data is published via a REST API and optionally to a MQTT broker. Support for Home Assistant autodiscovery is also provided in the MQTT connection.

# Requirements

The server runs on Linux, OSX and Windows and needs Python >= 3.6

## Hardware requirements
- Supported Hanazeder system (see next paragraph)
- Linux, OSX or Windows system
  - A Raspberry Pi (even a Zero or 1st gen) works fine
- A serial (RS232) connection to the pump. This can take one of two forms
  - The system is directly connected to the Hanazeder board, typically through a RS232 to USB adapter. Please note that I needed a nullmodem cable to get data to flow, so if nothing can be read buy a RS232 null-modem cable and plug between device and pump controller.
  - A RS232-to-TCP converter [something like this](https://cablematic.com/de/products/category/server-tcpip-rs232/)

## Supported Hanazeder systems
Supported systems are
- Hanazeder FP1 (untested)
- Hanazeder FP2 (untested)
- Hanazeder FP3 (untested)
- Hanazeder FP6 (untested)
- Hanazeder FP10
- Hanazeder SH series might also work, please test and let me know!

As you can see quite a few of the systems are untested. Since we do not write anything
into the pumps I don't expect any issues. It might show more sensors or outlets than the pumps actually support.

# Installation
Install from pip:
```
pip install hanazeder_server
```


# Starting

Read the help for all possible command line flags:
```
python -m hanazeder_server.mqtt --help
```

## Web server

A simple example of reading from a USB-to-RS232 adapter (`/dev/ttyUSB0`) and making it available as REST endpoint at http://localhost:5000/api/data run
```
python -m hanazeder_server.mqtt --serial-port/dev/ttyUSB0 --mqtt-server 192.168.1.1
```

This can be run in any ASGI compatible web server, more in the [Quart documentation on deploying](https://pgjones.gitlab.io/quart/tutorials/deployment.html).

## MQTT publishing
Please note that the MQTT client still contains the REST API mentioned above.

A simple example of reading from a USB-to-RS232 adapter (`/dev/ttyUSB0`) and posting the data to the MQTT broker at `192.168.1.1` is
```
python -m hanazeder_server.mqtt --serial-port/dev/ttyUSB0 --mqtt-server 192.168.1.1
```

A more complex example will read from a RS232-to-TCP converter at 192.168.1.2 port 3000 and post the data to the MQTT broker at 192.168.1.1, using the username MQTT-USER and password MQTT-PASSWORD to log into the broker. It will print debug information to stdout. Also add Home Assistant autodiscovery data.
```
python -m hanazeder_server.mqtt --address 192.168.1.2 --port 3000  --mqtt-user MQTT-USER --mqtt-password MQTT-PASSWORD --debug --mqtt-server 192.168.1.1 --home-assistant
```



CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.1.1 (2022-09-09)
### Fix
* Try to fix CI fetching ([`1772d16`](https://github.com/unverbraucht/hanazeder_server/commit/1772d16c142f6e3c38eb0bf710d5131ce02dc342))
* Try to fix CI ([`93d72bb`](https://github.com/unverbraucht/hanazeder_server/commit/93d72bb80aada33d79642def445cc66ae91449b8))
* Try to fix CI pipeline ([`9505ad6`](https://github.com/unverbraucht/hanazeder_server/commit/9505ad6587d24c7009d3f113a722c5816e38b32c))
* Also build and push docker image ([`099619a`](https://github.com/unverbraucht/hanazeder_server/commit/099619aff77418fd3f7ac1761203d5f5476dd9ec))

## v1.1.0 (2022-09-08)
### Features
* Also publish outlet / pump states
* Fixes to documentation
* Respect --home-assistant flag

## v1.0.0 (2022-09-08)
### Breaking
* Mark as stable
* Depend on hanazeder==1.2.1 which fixes transitive dependencies

## v0.1.1 (2022-07-26)
### Fix
* Show controller info in REST API. Init names ([`24c5750`](https://github.com/unverbraucht/hanazeder_server/commit/24c5750f98a7a1851fa7eeed6f1deaaff354239e))


hanazeder_server/BaseServer.py
from typing import List
from hanazeder.Hanazeder import HanazederFP, SENSOR_LABELS
from hanazeder.comm import InvalidHeaderException, ChecksumNotMatchingException

import asyncio

class BaseServer:
    running = True
    names: List[str] = [None] * 16
    conn: HanazederFP = None
    sensor_value = [None] * 16
    energy = [None] * 3

    def __init__(self, device_id, debug):
        self.device_id = device_id
        self.debug = debug

    async def connect(
                    self,
                    serial_port: str = None,
                    address: str = None,
                    port: int = 3000):
        self.conn = HanazederFP(debug=self.debug, request_timeout=2)
        await self.conn.open(
            serial_port=serial_port,
            address=address,
            port=port,
            timeout=0.5)
        await self.conn.read_information()
        print(f'Connected to {self.conn.device_type.name} with version {self.conn.version}')

    async def read_names_block(self):
        # Read label from fixed list
        configs = await self.conn.read_config_block(27, 15)
        for index, config_label in enumerate(configs):
            self.names.append(None)
            if config_label.value > 0 and \
                    config_label.value < len(SENSOR_LABELS):
                self.names[index] = SENSOR_LABELS[config_label.value]
            else:
                self.names[index] = await self.conn.read_sensor_name(index)

    async def publish_base(self):
        pass

    async def run_loop(self):
        read_tasks = [self.conn.read_energy()]
        read_tasks.append(self.conn.read_outlets())
        # Read all sensor values
        for sensor_idx in range(0, 15):
            # Skip unconnected sensors
            if self.names[sensor_idx] is None:
                continue
            if self.debug:
                print(f'Reading sensor {sensor_idx}')
            read_tasks.append(self.conn.read_sensor(sensor_idx))
        self.energy, self.outlet_states, *readings = await asyncio.gather(*read_tasks)
        if self.debug:
            print('Energy readings:')
            print(f'  Total   {self.energy[0]}')
            print(f'  Current {self.energy[1]}')
            print(f'  Impulse {self.energy[2]}')
        for sensor_idx in range(0, 15):
            # Skip unconnected sensors
            if self.names[sensor_idx] is not None:
                self.sensor_value[sensor_idx] = readings.pop(0)
        if self.debug:
            print('Outlet states:')
            print(self.outlet_states)

    def as_dict(self):
        sensors = []
        # Add all sensor values
        for sensor_idx in range(0, 15):
            # Skip unconnected sensors
            if self.names[sensor_idx] is None:
                continue
            sensors.append({
                    "name": self.names[sensor_idx],
                    "value": self.sensor_value[sensor_idx]
                })
        return {
            "sensors": sensors,
            "energy": self.energy[0],
            "power": self.energy[1],
            "impulses": self.energy[2],
            "outlets": self.outlet_states,
            "running": True,
            "controller": {
                "device": self.conn.device_type.name,
                "version": self.conn.version
            }
        }

    def close(self):
        if self.conn and self.conn.connection:
            if self.debug:
                print('Closed connection: ', self.conn.connection)
            self.conn.connection.close()


hanazeder_server/MqttClient.py
from asyncio_mqtt import Client, ProtocolVersion, Will
from copy import copy
import json

from .BaseServer import BaseServer


class MqttClient(BaseServer):
    def __init__(
                    self,
                    device_id,
                    mqtt_server,
                    mqtt_user=None,
                    mqtt_password=None,
                    mqtt_port: int = 1883,
                    debug=False,
                    home_assistant=False):
        super(MqttClient, self).__init__(device_id, debug)
        self.base_topic = f'hanazeder/{device_id}'

        will = Will(f'{self.base_topic}/state', payload='offline', retain=True)
        self.home_assistant = home_assistant

        self.mqttc = Client(
                    mqtt_server,
                    port=mqtt_port,
                    will=will,
                    client_id="hanazeder",
                    clean_session=True,
                    protocol=ProtocolVersion.V311,
                    transport="tcp",
                    username=mqtt_user,
                    password=mqtt_password)
    
    async def connect(
                    self,
                    serial_port: str = None,
                    address: str = None,
                    port: int = 3000):
        await super(MqttClient, self).connect(serial_port, address, port)
        connections = []
        if serial_port:
            connections.append(['serial', serial_port])
        else:
            connections.append(['ip', f'{address}:{port}'])
        # HA base config topic
        self.ha_base_config = {
            'availability': [{'topic': f'{self.base_topic}/state'}],
            'device': {
                'manufacturer': "Hanazeder",
                'connections': connections,
                'identifiers': [self.device_id],
                'name': self.conn.device_type.name,
                'model': self.conn.device_type.name,
                'sw_version': self.conn.version,
            },
            'enabled_by_default': True
        }
        await self.mqttc.connect()

        await self.mqttc.publish(
            f'{self.base_topic}/state', 'online',
            retain=True)

    async def publish_base(self):
        await super(MqttClient, self).publish_base()
        if self.home_assistant:
            await self.publish_home_assistant_autodiscovery()
    
    async def publish_home_assistant_autodiscovery(self):
        for sensor_idx in range(0, 15):
            name = self.names[sensor_idx]
            print(f'Sensor {sensor_idx} has name {name}')
            unique_id = f'{self.device_id}-f{sensor_idx}'
            ha_config = copy(self.ha_base_config)
            ha_config['device_class'] = 'temperature'
            ha_config['state_class'] = 'measurement'
            ha_config['name'] = name
            ha_config['unit_of_measurement'] = '°C'
            ha_config['unique_id'] = unique_id
            ha_config['state_topic'] = f'{self.base_topic}/sensor/{sensor_idx}'
            ha_config['value_template'] = '{{ value_json.temperature }}'
            await self.mqttc.publish(
                f'homeassistant/sensor/{unique_id}/temperature/config',
                json.dumps(ha_config, ensure_ascii=False).encode('utf8'),
                retain=True)

        for sensor_idx in range(0, 10):
            unique_id = f'{self.device_id}-o{sensor_idx}'
            ha_config = copy(self.ha_base_config)
            ha_config['unique_id'] = unique_id
            ha_config['name'] = f'Pump {sensor_idx + 1}'
            ha_config['state_topic'] = f'{self.base_topic}/outlet/{sensor_idx}'
            ha_config['value_template'] = '{{ value_json.state }}'
            await self.mqttc.publish(
                f'homeassistant/binary_sensor/{unique_id}/outlet/config',
                json.dumps(ha_config, ensure_ascii=False).encode('utf8'),
                retain=True)

        # Setup energy config
        ha_config = copy(self.ha_base_config)
        unique_id = f'{self.device_id}-power'
        ha_config['device_class'] = 'power'
        ha_config['state_class'] = 'measurement'
        ha_config['name'] = f'{self.device_id} Power'
        ha_config['unit_of_measurement'] = 'kW'
        ha_config['unique_id'] = unique_id
        ha_config['state_topic'] = f'{self.base_topic}/power'
        await self.mqttc.publish(
            f'homeassistant/sensor/{unique_id}/power/config',
            json.dumps(ha_config, ensure_ascii=False).encode('utf8'),
            retain=True)

        ha_config = copy(self.ha_base_config)
        unique_id = f'{self.device_id}-energy'
        ha_config['device_class'] = 'energy'
        ha_config['state_class'] = 'total'
        ha_config['name'] = f'{self.device_id} Energy'
        ha_config['unit_of_measurement'] = 'kWh'
        ha_config['unique_id'] = unique_id
        ha_config['state_topic'] = f'{self.base_topic}/energy'
        await self.mqttc.publish(
            f'homeassistant/sensor/{unique_id}/energy/config',
            json.dumps(ha_config, ensure_ascii=False).encode('utf8'),
            retain=True)

        ha_config = copy(self.ha_base_config)
        unique_id = f'{self.device_id}-impulse'
        ha_config['device_class'] = 'frequency'
        ha_config['state_class'] = 'measurement'
        ha_config['name'] = f'{self.device_id} Impulses'
        ha_config['unit_of_measurement'] = 'Hz'
        ha_config['unique_id'] = unique_id
        ha_config['state_topic'] = f'{self.base_topic}/impulse'
        await self.mqttc.publish(
            f'homeassistant/sensor/{unique_id}/impulse/config',
            json.dumps(ha_config, ensure_ascii=False).encode('utf8'),
            retain=True)
    
    async def run_loop(self):
        await super(MqttClient, self).run_loop()
        # Read all sensor values
        for sensor_idx in range(0, 15):
            # Skip unconnected sensors
            if \
                    self.names[sensor_idx] is None \
                    or self.sensor_value[sensor_idx] is None:
                continue
            await self.mqttc.publish(
                f'{self.base_topic}/sensor/{sensor_idx}',
                json.dumps({'temperature': self.sensor_value[sensor_idx]}))
        
        for i, state in enumerate(self.outlet_states):
            await self.mqttc.publish(
                f'{self.base_topic}/outlet/{i}',
                json.dumps({'state': 'ON' if state else 'OFF'}))

        # TODO: parallelize
        await self.mqttc.publish(f'{self.base_topic}/energy', self.energy[0])
        await self.mqttc.publish(f'{self.base_topic}/power', self.energy[1])
        await self.mqttc.publish(f'{self.base_topic}/impulse', self.energy[2])

    def shutdown(self):
        self.running = False


hanazeder_server/mqtt.py
import asyncio
import argparse
import sys
import logging

from quart import Quart

from hanazeder_server.BaseServer import BaseServer
from hanazeder_server.MqttClient import MqttClient

logging.basicConfig(format='%(asctime)s %(message)s')

parser = argparse.ArgumentParser()
parser.add_argument("--serial-port", help="set serial port",
                    type=str)
parser.add_argument("--mqtt-server", help="MQTT server")
parser.add_argument("--mqtt-port", help="MQTT server port", default=1883,
                    type=int)
parser.add_argument("--mqtt-user", help="MQTT username", type=str)
parser.add_argument("--mqtt-password", help="MQTT password", type=str,
                    default=None)
parser.add_argument(
    "--home-assistant",
    help="Set Home Assistant auto-detect",
    action="store_true")
parser.add_argument(
    "--device-id",
    help="Device id, used for topic structure and to identify within HA",
    default="hanazeder")
parser.add_argument("--debug", help="print low-level messages", action="store_true")
parser.add_argument(
    "--address",
    help="connect to HOSTNAME RS232-adapter, needs port as well",
    type=str)
parser.add_argument(
    "--port",
    help="connect to HOSTNAME on port PORT",
    type=int, default=5000)
args = parser.parse_args()

if args.address and args.serial_port:
    print('Cannot specify both serial-port and address')
    sys.exit(1)
if not args.address and not args.serial_port:
    print('Specify either serial port or RS-232 converter hostname')
    sys.exit(1)

if args.address and not args.port:
    print('Specify port together with address')
    sys.exit(2)

if args.debug:
    logging.basicConfig(level=logging.DEBUG)

app = Quart(__name__)


def create_instance():
    if args.mqtt_server:
        return MqttClient(
                    args.device_id,
                    args.mqtt_server,
                    args.mqtt_user,
                    args.mqtt_password,
                    args.mqtt_port,
                    args.debug,
                    args.home_assistant)
    else:
        return BaseServer(args.device_id, args.debug)


async def mqtt_loop():
    while not hasattr(app, 'mqtt_instance') or app.mqtt_instance.running:
        try:
            app.mqtt_instance = create_instance()
            await app.mqtt_instance.connect(
                args.serial_port,
                args.address,
                args.port)
            await app.mqtt_instance.read_names_block()
            await app.mqtt_instance.publish_base()
            while app.mqtt_instance.running:
                await app.mqtt_instance.run_loop()
                if not app.mqtt_instance.conn.connected:
                    break
                await asyncio.sleep(30)
        except Exception as e:
            try:
                app.mqtt_instance.close()
            except Exception as e2:
                print('Cannot close connection', e2)
            print('Error while reading, will sleep and retry', e)
            if app.mqtt_instance.running:
                await asyncio.sleep(90)


@app.before_serving
async def startup():
    asyncio.get_event_loop().create_task(mqtt_loop())


@app.after_serving
async def shutdown():
    if args.mqtt_server:
        app.mqtt_instance.shutdown()
    app.mqtt_instance.close()


@app.get("/api/data")
async def echo():
    if app.mqtt_instance:
        return app.mqtt_instance.as_dict()
    return {
        "running": False
    }

app.run()
